File: packages/lynx-websocket/ios/WebSocketEventBus.swift
import Foundation

/// Process-wide pub/sub bus that owns the conversion of native WebSocket
/// callbacks into JS-side event payloads. Sockets, sessions, and OS
/// delegates write here; per-LynxView `WebSocketPublisher` instances read.
///
/// Mirrors the publisher pattern used elsewhere in the repo
/// (`LinkingState`, `SafeAreaPublisher`, …): native lifecycle is decoupled
/// from any specific LynxView so events survive view recreation and
/// per-view subscribers can fan out without holding onto the WS task.
final class WebSocketEventBus {

    static let shared = WebSocketEventBus()

    /// JSON-encodable payload pushed to JS as the single
    /// `__sigxWebSocketEvent` global event. Keys match the `NativeEvent`
    /// interface in `src/websocket.ts`.
    typealias Payload = [String: Any]
    typealias Listener = (Payload) -> Void

    private let queue = DispatchQueue(label: "com.sigx.websocket.bus")
    private var listeners: [(token: UUID, fn: Listener)] = []

    @discardableResult
    func addListener(_ fn: @escaping Listener) -> UUID {
        let token = UUID()
        queue.sync { listeners.append((token, fn)) }
        return token
    }

    func removeListener(_ token: UUID) {
        queue.sync { listeners.removeAll { $0.token == token } }
    }

    // MARK: - Publish

    func publish(open protocolStr: String, extensions: String, id: Int) {
        emit([
            "id": id,
            "type": "open",
            "protocol": protocolStr,
            "extensions": extensions,
        ])
    }

    func publish(messageText text: String, id: Int) {
        emit([
            "id": id,
            "type": "message",
            "data": text,
            "isBinary": false,
        ])
    }

    func publish(messageBinary base64: String, id: Int) {
        emit([
            "id": id,
            "type": "message",
            "binary": base64,
            "isBinary": true,
        ])
    }

    func publish(error message: String, id: Int) {
        emit([
            "id": id,
            "type": "error",
            "data": message,
        ])
    }

    func publish(close code: Int, reason: String, wasClean: Bool, id: Int) {
        emit([
            "id": id,
            "type": "close",
            "code": code,
            "reason": reason,
            "wasClean": wasClean,
        ])
    }

    private func emit(_ payload: Payload) {
        let snapshot = queue.sync { listeners }
        for (_, fn) in snapshot { fn(payload) }
    }
}

File: packages/lynx-websocket/android/com/sigx/websocket/WebSocketEventBus.kt
package com.sigx.websocket

import com.lynx.react.bridge.JavaOnlyMap
import java.util.UUID
import java.util.concurrent.CopyOnWriteArrayList

/**
 * Process-wide pub/sub bus that owns the conversion of native WebSocket
 * callbacks into JS-side event payloads. OkHttp listeners write here;
 * per-LynxView [WebSocketPublisher] instances read.
 *
 * Mirrors the publisher pattern used elsewhere in the repo
 * (`LinkingState`, `SafeAreaPublisher`, …): native lifecycle is decoupled
 * from any specific LynxView so events survive view recreation and
 * per-view subscribers can fan out without holding onto the WS task.
 *
 * Payload keys match the `NativeEvent` interface in `src/websocket.ts`.
 */
internal object WebSocketEventBus {

    private val listeners = CopyOnWriteArrayList<Pair<UUID, (JavaOnlyMap) -> Unit>>()

    fun addListener(fn: (JavaOnlyMap) -> Unit): UUID {
        val token = UUID.randomUUID()
        listeners.add(token to fn)
        return token
    }

    fun removeListener(token: UUID) {
        listeners.removeAll { it.first == token }
    }

    fun publishOpen(id: Int, protocol: String, extensions: String) {
        emit(
            mapOf(
                "id" to id,
                "type" to "open",
                "protocol" to protocol,
                "extensions" to extensions,
            )
        )
    }

    fun publishMessageText(id: Int, text: String) {
        emit(
            mapOf(
                "id" to id,
                "type" to "message",
                "data" to text,
                "isBinary" to false,
            )
        )
    }

    fun publishMessageBinary(id: Int, base64: String) {
        emit(
            mapOf(
                "id" to id,
                "type" to "message",
                "binary" to base64,
                "isBinary" to true,
            )
        )
    }

    fun publishError(id: Int, message: String) {
        emit(
            mapOf(
                "id" to id,
                "type" to "error",
                "data" to message,
            )
        )
    }

    fun publishClose(id: Int, code: Int, reason: String, wasClean: Boolean) {
        emit(
            mapOf(
                "id" to id,
                "type" to "close",
                "code" to code,
                "reason" to reason,
                "wasClean" to wasClean,
            )
        )
    }

    private fun emit(payload: Map<String, Any>) {
        // JavaOnlyMap doesn't ship a `from(Map)` overload that handles
        // mixed value types on every host, so build it explicitly to keep
        // the bridge happy.
        val map = JavaOnlyMap()
        for ((k, v) in payload) {
            when (v) {
                is String -> map.putString(k, v)
                is Int -> map.putInt(k, v)
                is Double -> map.putDouble(k, v)
                is Boolean -> map.putBoolean(k, v)
                else -> map.putString(k, v.toString())
            }
        }
        for ((_, fn) in listeners) fn(map)
    }
}
